src/metabase/driver/sql/query_processor.clj

```diff
@@ -68,6 +68,18 @@
   additional keys there and need to access them."
   nil)
 
+(def ^:dynamic *parent-expr-type-info*
+  "Bound to the type info (i.e. result of calling (hx/type-info expr)) of a parent expression. Useful if a child
+  expression needs to vary its behavior according to the parent's type. Rebind as necessary when expanding complex
+  expressions.
+
+  Example: in Presto, checkins.date AT TIME ZONE 'America/Los_Angeles' is not allowed, because that column is of
+  type 'date', and the 'AT TIME ZONE' operator can only be invoked on types 'timestamp with time zone' and 'time
+  with time zone'. However, if a date bucketing expression is being expanded, the innermost ->honeysql form may
+  not have access to the original column's type info (since it's actually inside a SQL call to add or subtract dates,
+  etc."
+  nil)
+
```

Comment:
It seems like this is going to break as soon as the expression ends up in a situation where the _grandparent_ has the type info rather than the parent. It's also extremely misleading to document this variable as being `Bound to the type info [...] of a parent expression` when it's actually only bound that way in one single very specific situation. It's also a little misleading in the sense that the `expr` that gets passed to `(date :week)` isn't really what I'd call the "parent expression" if it's getting wrapped and passed to other methods as `expr`... I don't expect `*parent-expr-type-info*` to mean "the type info of the arg you're passed before it got wrapped".

It seems sort of weird that you'd need to know the parent expression type information anyway in order to do something like apply `week-of-year` bucketing to a date column. I'm guessing you're doing it so the `:day-of-year` method sees it. Binding a dynamic variable in order to pass along type information about the expression we've wrapped seems like the wrong way to go about it. Why not just have the wrapped expression include type information?

Just make sure the `:week` implementation returns a HoneySQL form with type information. Arguably all the 
`sql.qp/date` truncation methods should be returning HoneySQL forms with type info. Any sort of truncation operation that returns another temporal type is returning _some_ type of value, maybe a `DATE` or `DATETIME` or `TIMESTAMP` or `TIME` or something else. We should pass that information along so we don't have to resort to weird dynamic variable hacks.

(That's the whole point of having the `hx/with-type-info` stuff in the first place -- so we can pass along information about what type a HoneySQL expression produces so we can do the appropriate thing in places where we need different functions/casts for different types without having to bind a bunch of dynamic variables that make the logic hard to follow.)

Reply:
Fair point.  Let me step back and explain the problem and maybe you can suggest a better solution.

Here is the problematic expression:
```
(hsql/call
 :date_add
 (literal "day")
 -1
 (hsql/call
  :date_trunc
  (literal "week")
  (hsql/call
   :date_add
   (literal "day")
   1
   (metabase.util.honeysql-extensions/with-type-info
    (hx/identifier :field "default" "checkins" "date")
    #:metabase.util.honeysql-extensions{:database-type "date"}))))
```

This is basically part of a `:day-of-week` truncation (which of course comes from `(defmethod date [:sql :week-of-year]...` in the QP itself).

Now, the current code in Presto JDBC (which you originally helped write) is basically using this `AtTimeZone` mechanism to add the `AT TIME ZONE` bit when the report timezone is defined.  The trouble is, this type of bucketing can very well be applied to both `date`, `timestamp`, and `timestamp with time zone` columns.   **Only** in the latter two cases, should this expression be added to the native query.  If we add it when the "underlying" (sorry, lack of a better term for it) column is a type `date` (as in the example above), then this will fail with:

```Type of value must be a time or timestamp with or without time zone (actual date)```

So, the task at hand is to **suppress** this extra expression in the case of date.  I attempted to do this poorly through the dynamic var.  Now, ripping that out as per your feedback... what next?  Should the `AtTimeZone` function walk the entire expression tree searching for the "underlying" base type (which, as can be seen, is available in some fashion).

Reply:
OK, I think I figured out how to deal with it properly: basically by switching from an exclusion list to an inclusion one.  In the `AtTimeZone` handling, I will only set the expression if the type is KNOWN to be a valid one (`time`, `time with time zone`, `timestamp`, or `timestamp with time zone`).  That way, the innermost call (i.e. the column itself) will be handled correctly, and the wrapping function calls don't need to care about it (and the Presto query engine presumably should just do the right thing in that case).